### PR TITLE
[Lok Qi Jun] iP

### PR DESCRIPTION
**Pull request by LokQiJun**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,120 @@
+import java.util.*;
```
Comment (xzynos) [inline src/main/java/Duke.java]: The coding standard states not to use wildcard imports

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,120 @@
...
 public class Duke {
+    private static ArrayList<Task> lists = new ArrayList<>();
     public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
+        String inputs;
+        String[] splitInputs;
+        doGreetings();
+        inputs = scanner.nextLine();
+        while(!(inputs.equalsIgnoreCase("bye"))) {
+            System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
```
Comment (xzynos) [inline src/main/java/Duke.java]: This magic string does not provide any context about why it is being printed

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,120 @@
...
+                }
+                else if(Integer.parseInt(splitInputs[1]) > lists.size()){
+                    System.out.println("There is no task with index " + splitInputs[1]);
+                }
+                else {
+                    lists.get(Integer.parseInt(splitInputs[1]) - 1).setDone(false);
+                    doPrintlists();
+                }
+            }
+            else if(splitInputs[0].equalsIgnoreCase("todo") || splitInputs[0].equalsIgnoreCase("event") || splitInputs[0].equalsIgnoreCase("deadline")){
```
Comment (xzynos) [inline src/main/java/Duke.java]: This expression is complicated and hard to read

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,120 @@
+import java.util.*;
 public class Duke {
```
Comment (junhaoliu2468) [inline src/main/java/Duke.java]: can consider using more classes - will make things much simpler to work with in the later stages

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,120 @@
...
+                }
+            }
+            else if(splitInputs[0].equalsIgnoreCase("todo") || splitInputs[0].equalsIgnoreCase("event") || splitInputs[0].equalsIgnoreCase("deadline")){
+                lists.add(new Task(splitInputs));
+                System.out.println("\t Added: ");
+                lists.get(lists.size() - 1).doPrint();
+                System.out.println("\t There are currently " + lists.size() + " task(s) in the list :)");
+            }
+            else {
+                System.out.println("Sorry I do not understand your command :(");
```
Comment (junhaoliu2468) [inline src/main/java/Duke.java]: i think idea is along the same line - you have different objects so you have more calls - it helps to make your code much less cluttered and you can make use of super() functions to reduce duplicate code too

File: src/main/java/Duke.java [hunk 6]
```diff
@@ -1,10 +1,120 @@
...
+        protected char taskType;
+
+        protected boolean isDone;
+
+        public Task(String[] splitInputs) {
+            int i = 1;
+            this.taskType = Character.toUpperCase(splitInputs[0].charAt(0));
+            while(i < splitInputs.length && !(splitInputs[i].contains("at") || splitInputs[i].contains("by"))){
+                this.description += splitInputs[i++];
+                this.description += " ";
```
Comment (junhaoliu2468) [inline src/main/java/Duke.java]: very good effort honestly in splitting up the cases

File: src/main/java/Duke.java [hunk 7]
```diff
@@ -1,10 +1,120 @@
...
+            if(taskType == 'D') {
+                System.out.print("by : ");
+            }
+            if(taskType == 'E'){
+                System.out.print("at : ");
+            }
+            System.out.println(deadline);
+        }
+    }
+
```
Comment (junhaoliu2468) [inline src/main/java/Duke.java]: can again consider looking at things from a more macro - object kind of perspective

File: src/main/java/Duke.java [hunk 8]
```diff
@@ -1,10 +1,120 @@
...
+            }
+            else if(splitInputs[0].equalsIgnoreCase("todo") || splitInputs[0].equalsIgnoreCase("event") || splitInputs[0].equalsIgnoreCase("deadline")){
+                lists.add(new Task(splitInputs));
+                System.out.println("\t Added: ");
+                lists.get(lists.size() - 1).doPrint();
+                System.out.println("\t There are currently " + lists.size() + " task(s) in the list :)");
+            }
+            else {
+                System.out.println("Sorry I do not understand your command :(");
+            }
```
Comment (xzynos) [inline src/main/java/Duke.java]: This code block is hard to follow. Do consider applying the Single Level of Abstraction Principle (SLAP)

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -1,10 +1,120 @@
...
+            System.out.print(i+1 + ": ");
+            lists.get(i).doPrint();
+        }
+    }
+
+    public static void doExit() {
+        System.out.println(" Bye! Hope to see you again soon :)");
+    }
+
+    public static void doGreetings(){
```
Comment (xzynos) [inline src/main/java/Duke.java]: The coding standard states that a space is required between the closing bracket ) and the opening parenthesis {

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -1,10 +1,120 @@
...
         String logo = " ____        _        \n"
                 + "|  _ \\ _   _| | _____ \n"
                 + "| | | | | | | |/ / _ \\\n"
                 + "| |_| | |_| |   <  __/\n"
                 + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        System.out.println("Hello from\n" + logo + "\n \n Hello! I'm Duke\n What can I do for you?");
     }
```
Comment (xzynos) [inline src/main/java/Duke.java]: The naming of the functions are clear and is easy to understand their purpose

**Review by xzynos — COMMENTED** [review]
This is a good attempt but do consider refactoring the code to make it easier to follow

File: src/main/java/Duke.java [hunk 11]
```diff
@@ -1,10 +1,120 @@
...
+                System.out.print("by : ");
+            }
+            if(taskType == 'E'){
+                System.out.print("at : ");
+            }
+            System.out.println(deadline);
+        }
+    }
+
+    public static void doPrintlists() {
```
Comment (lcsroy) [inline src/main/java/Duke.java]: the naming convention is well used

**Review by lcsroy — COMMENTED** [review]
correct naming convention and "{}" are standardised across all if-else statement

File: src/main/java/Duke/Deadline.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+        i += 1;
+        if(i >= splitInputs.length) {
+            throw new DukeException();
+        }
+        super.dueDate = "( by :";
+        while(i < splitInputs.length) {
+            super.dueDate += splitInputs[i++];
+            super.dueDate += " ";
+        }
+        super.dueDate += ")";
```
Comment (okkhoy) [inline src/main/java/Duke/Deadline.java]: consider a more intuitive way of handling parameters/iterating

File: src/main/java/Duke/ExceptionHandler.java [hunk 13]
```diff
@@ -0,0 +1,16 @@
+package Duke;
+
+public class ExceptionHandler {
```
Comment (okkhoy) [inline src/main/java/Duke/ExceptionHandler.java]: Possibly rename? You are not doing any exception handling here. This class simply provides error messages, so you can name it as such

File: src/main/java/Duke/TaskManager.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+    public void unmarkDone(int taskNumber) {
+        tasks.get(taskNumber).setDone(false);
+        System.out.println("Nice! I've marked this task as done:");
+        System.out.println();
+    }
+    public void printList() throws DukeException{
+        if(tasks.size() == 0) {
+            throw new DukeException();
+        } else {
```
Comment (okkhoy) [inline src/main/java/Duke/TaskManager.java]: Use this as guard clause and eliminate the tiny arrowhead code below

File: src/main/java/Duke/UserInterface.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+    public UserInterface() {
+        scanner = new Scanner(System.in);
+        myTaskManager = new TaskManager();
+    }
+
+    public void runProgram() {
+        String inputs;
+        inputs = scanner.nextLine();
+        while(!(inputs.equalsIgnoreCase(COMMAND_BYE))) {
```
Comment (okkhoy) [inline src/main/java/Duke/UserInterface.java]: It would be better to gather the command word in an explicit variable and use it instead of indexed array element.

File: src/main/java/Duke/UserInterface.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+                    myTaskManager.addTasks(new Deadline(splitInputs));
+                    System.out.println("\t Added: ");
+                    myTaskManager.printTask(myTaskManager.getSize() - 1);
+                    myTaskManager.printSize();
+                } catch(DukeException e) {
+                    System.out.println(ExceptionHandler.DEADLINE_INPUT_ERROR);
+                }
+            }
+            else {
+                System.out.println(ExceptionHandler.UNKNOWN_INPUTS);
```
Comment (okkhoy) [inline src/main/java/Duke/UserInterface.java]: Coding standard violation present here.

File: src/main/java/Duke/Deadline.java [hunk 17]
```diff
@@ -0,0 +1,23 @@
+package Duke;
+public class Deadline extends Task{
```
Comment (okkhoy) [inline src/main/java/Duke/Deadline.java]: Spacing is a bit off in many files. You can get it fixed using the IDE!

**Review by okkhoy — COMMENTED** [review]
In general, the code is quite well written.
You can improve the spacing in some of the files.
